feelpp/feel/feelmesh/meshsupport.hpp

```diff
@@ -363,18 +347,124 @@ private :
+    {
+        for ( auto const& eltWrap : M_rangeElements )
+            myExtendedElements->push_back( eltWrap );
+    }
+
+    if ( ( ( entity == EntityProcessType::GHOST_ONLY ) || ( entity == EntityProcessType::ALL ) ) && ( M_mesh->worldComm().localSize() > 1 ) )
+    {
+        CHECK( M_hasUpdatedParallelData ) << "parallel data must be updated";
+
+        std::unordered_set<size_type> eltGhostDone;
+        for ( auto const& faceWrap : M_rangeInterProcessFaces )
+        {
+            auto const& faceip = boost::unwrap_ref( faceWrap ); //*face_it );
+            auto const& elt0 = faceip.element0();
+            auto const& elt1 = faceip.element1();
+            const bool elt0isGhost = elt0.isGhostCell();
+            auto const& eltOffProc = ( elt0isGhost ) ? elt0 : elt1;
+            auto const& eltOnProc = ( elt0isGhost ) ? elt1 : elt0;
+            if ( eltGhostDone.find( eltOffProc.id() ) != eltGhostDone.end() )
+                continue;
+            myExtendedElements->push_back( boost::cref( eltOffProc ) );
+            eltGhostDone.insert( eltOffProc.id() );
+        }
+    }
+    range_elements_type rangeExtendedElements = boost::make_tuple( mpl::size_t<MESH_ELEMENTS>(), myExtendedElements->begin(), myExtendedElements->end(), myExtendedElements );
+    return rangeExtendedElements;
+}
+
+template<typename MeshType>
+typename MeshSupport<MeshType>::range_faces_type const&
+MeshSupport<MeshType>::rangeMarkedFaces( uint16_type marker_t, boost::any flag )
+{
+    std::set<flag_type> markerFlagSet = Feel::unwrap_ptr( M_mesh ).markersId( flag );
+    flag_type m = *markerFlagSet.begin();
+    if ( M_isFullSupport )
+    {
+        M_rangeMarkedFaces[m] = markedfacesByType( M_mesh, marker_t, flag );
+        return M_rangeMarkedFaces.at(m);
+    }
+        
+    
+    if ( M_rangeMarkedFaces.count( m ) )
+        return M_rangeMarkedFaces.at( m );
+    typename MeshTraits<mesh_type>::faces_reference_wrapper_ptrtype myfaces( new typename MeshTraits<mesh_type>::faces_reference_wrapper_type );
+    for ( auto const& eltWrap : this->rangeElements() )
+    {
+        auto const& elt = unwrap_ref( eltWrap );
+        for ( uint16_type i = 0; i < mesh_type::element_type::numTopologicalFaces; ++i )
+        {
+            if ( !elt.facePtr(i) )
+                continue;
+            const face_type* facePtr = elt.facePtr(i);
+            size_type faceId = facePtr->id();
+            auto const& face = elt.face(i);
+            if ( !face.hasMarker( marker_t ) )
+                continue;
+            if ( face.marker( marker_t ).isOff() )
+                continue;
+            if ( markerFlagSet.find( face.marker( marker_t ).value() ) == markerFlagSet.end() )
+                continue;
+            
+            myfaces->push_back( boost::cref( face ) );
+        }
+    }
+    M_rangeMarkedFaces[m] = boost::make_tuple( mpl::size_t<MESH_FACES>(),myfaces->begin(),myfaces->end(),myfaces );
+    return M_rangeMarkedFaces.at( m );
+}
```

Comment:
@vincentchabannes so you mean I shouldn't store in  `M_rangeMarkedFaces` and just return 
```
boost::make_tuple( mpl::size_t<MESH_FACES>(),myfaces->begin(),myfaces->end(),myfaces );
```

Reply:
and this should be recomputed all the time it is called?

Reply:
not exactly, you should use the internal/boundary range of faces, it should be more efficient that use the element of range

Reply:
Ah ok I understand